### ISSUE TITLE
Questions about fungible asset transfer

### ISSUE THREAD

**Issue by claudiosdc** [rgb-org/rgb-node#128]
I am very interested in the work that you guys are doing, so I have been doing as much research on it as possible. As part of this research, I have run a simple test (using version 0.2.1 of RGB node) to issue and transfer fungible assets.

The test involved two RGB nodes (`rgb-cli` and `rgb-cli2`), and it was comprised of the following steps:

1. Issuance of a new fungible asset.

```shell
rgb-cli2 fungible issue -d "Test RGB asset #2" -p 2 XYZB "XYZ Test 2" 3000@8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:9
```

2. Retrieval of asset ID.

```shell
rgb-cli2 fungible list
```

output:

```shell
---
- name: XYZ Test 2
  id: rgb1zsetldhesgr2cjwaxunv8vfl4g3ln3f8zpv22rlm75cgklv94hpqu78w0t
  ticker: XYZB
```

3. Sharing of newly created asset.

```shell
rgb-cli2 fungible export rgb1zsetldhesgr2cjwaxunv8vfl4g3ln3f8zpv22rlm75cgklv94hpqu78w0t
```

output:

```shell
Asset successfully exported. Use this information for sharing:
genesis1qyfe883hey6jrgj2xvk5g3dfmfqfzm7a4wez4pd2krf7ltsxffd6u6nrvjvvnc8vt9llmp7663pgututl9heuwaudet72ay9j6thc6cetuvhxvsqqya5xjt2w9y4u6sfkuszwwctnrpug5yjxnthmr3mydg05rdrpspcxysnqvvqpfvag2w8jxzzsz9pf8pjfwf0xvln5z7w93yjln3gcnyxsa04jsf2p8vu4sxgeqrzyxg5tyvzyg4rnsvyznpwzpfszmyuzyf5ssqsvfmfpwungf38zut23xptyyfn2zwgrzg9fper83l3yjqgcksquscwlcyx8zgezcfpqv3n8j9crt26t04jhuveekvlwxlgm9hgl0c7k4s45vwy0txle44staexvc7hnu5sahx4un5xpx7jpa7qk54nqyqq5qcxzqqqxff36e
```

then on the other node:

```shell
rgb-cli fungible import genesis1qyfe883hey6jrgj2xvk5g3dfmfqfzm7a4wez4pd2krf7ltsxffd6u6nrvjvvnc8vt9llmp7663pgututl9heuwaudet72ay9j6thc6cetuvhxvsqqya5xjt2w9y4u6sfkuszwwctnrpug5yjxnthmr3mydg05rdrpspcxysnqvvqpfvag2w8jxzzsz9pf8pjfwf0xvln5z7w93yjln3gcnyxsa04jsf2p8vu4sxgeqrzyxg5tyvzyg4rnsvyznpwzpfszmyuzyf5ssqsvfmfpwungf38zut23xptyyfn2zwgrzg9fper83l3yjqgcksquscwlcyx8zgezcfpqv3n8j9crt26t04jhuveekvlwxlgm9hgl0c7k4s45vwy0txle44staexvc7hnu5sahx4un5xpx7jpa7qk54nqyqq5qcxzqqqxff36e
```

output:

```shell
Asset successfully imported
```

4. Creation of invoice to receive asset.

```shell
rgb-cli fungible invoice rgb1zsetldhesgr2cjwaxunv8vfl4g3ln3f8zpv22rlm75cgklv94hpqu78w0t 30 d092aaaed065bee333d68f1bf67a9893c09428da957e1f7ccee28a2f2699b764:3
```

output:

```shell
Invoice: rgb20:utxob1zh2c4ygwds99ma6n02k22asxgf8yl7k4mf574vd06s7ta0r9gz4qzr306r?asset=rgb1zsetldhesgr2cjwaxunv8vfl4g3ln3f8zpv22rlm75cgklv94hpqu78w0t&amount=30
Outpoint blinding factor: 18445135181747981829
```

5. Transfer of asset.

```shell
rgb-cli2 fungible transfer 'rgb20:utxob1zh2c4ygwds99ma6n02k22asxgf8yl7k4mf574vd06s7ta0r9gz4qzr306r?asset=rgb1zsetldhesgr2cjwaxunv8vfl4g3ln3f8zpv22rlm75cgklv94hpqu78w0t&amount=30' /var/lib/rgb/transf_tx3.psbt /var/lib/rgb/consignment3.rgb /var/lib/rgb/witness3.psbt -i 8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:9 -a 2970@8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:0
```

output:

```shell
Transfer succeeded, consignment data are written to "/var/lib/rgb/consignment3.rgb", partially signed witness transaction to "/var/lib/rgb/witness3.psbt"
```

6. Finalization of remittance of witness transaction.

7. Validation and acceptance of asset transfer.

```shell
rgb-cli fungible validate /var/lib/rgb/consignment3.rgb
```

output:

```shell
Asset transfer successfully validated.
```

then:

```shell
rgb-cli fungible accept /var/lib/rgb/consignment3.rgb d092aaaed065bee333d68f1bf67a9893c09428da957e1f7ccee28a2f2699b764:3 18445135181747981829
```

output:

```shell
Asset transfer successfully accepted.
```

At the end of the test, I have checked the status of the asset on each node, and I got the following.

RGB node 1 (`rgb-cli fungible list -l`):

```shell
---
- id: rgb1zsetldhesgr2cjwaxunv8vfl4g3ln3f8zpv22rlm75cgklv94hpqu78w0t
  ticker: XYZB
  name: XYZ Test 2
  description: "Test RGB asset #2"
  supply:
    known_circulating:
      - 3000
      - 2
    is_issued_known: ~
    max_cap:
      - 3000
      - 2
  chain: testnet3
  fractional_bits: 2
  date: "2021-01-23T17:33:55"
  known_issues:
    - id: c2ad857d8b30f5fb0fa5581027c5f923aa3fb1c32637dd49ac0682f9b6bf3214
      asset_id: rgb1zsetldhesgr2cjwaxunv8vfl4g3ln3f8zpv22rlm75cgklv94hpqu78w0t
      amount:
        - 3000
        - 2
      origin: ~
  known_inflation: {}
  unknown_inflation:
    - 0
    - 0
  known_allocations:
    "8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:9":
      - node_id: c2ad857d8b30f5fb0fa5581027c5f923aa3fb1c32637dd49ac0682f9b6bf3214
        index: 0
        outpoint: "8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:9"
        value:
          value: 3000
          blinding: "0000000000000000000000000000000000000000000000000000000000000001"
    "d092aaaed065bee333d68f1bf67a9893c09428da957e1f7ccee28a2f2699b764:3":
      - node_id: d6a6dbc8a1d249111f8844fcd4d0b15fbc7017705cac29bbd9929eaa63e16d75
        index: 1
        outpoint: "d092aaaed065bee333d68f1bf67a9893c09428da957e1f7ccee28a2f2699b764:3"
        value:
          value: 30
          blinding: 2d7b929dfa04dcbeb9db18273ce3a60633e26a5355b610c4fb5de0fa9a99a373
```

RGB node 2 (`rgb-cli2 fungible list -l`):

```shell
---
- id: rgb1zsetldhesgr2cjwaxunv8vfl4g3ln3f8zpv22rlm75cgklv94hpqu78w0t
  ticker: XYZB
  name: XYZ Test 2
  description: "Test RGB asset #2"
  supply:
    known_circulating:
      - 3000
      - 2
    is_issued_known: ~
    max_cap:
      - 3000
      - 2
  chain: testnet3
  fractional_bits: 2
  date: "2021-01-23T17:33:55"
  known_issues:
    - id: c2ad857d8b30f5fb0fa5581027c5f923aa3fb1c32637dd49ac0682f9b6bf3214
      asset_id: rgb1zsetldhesgr2cjwaxunv8vfl4g3ln3f8zpv22rlm75cgklv94hpqu78w0t
      amount:
        - 3000
        - 2
      origin: ~
  known_inflation: {}
  unknown_inflation:
    - 0
    - 0
  known_allocations:
    "8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:9":
      - node_id: c2ad857d8b30f5fb0fa5581027c5f923aa3fb1c32637dd49ac0682f9b6bf3214
        index: 0
        outpoint: "8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:9"
        value:
          value: 3000
          blinding: "0000000000000000000000000000000000000000000000000000000000000001"
```

Based on that result, I have the following questions:

1. I was assuming that the PBST that is passed as a parameter to the `fungible transfer` command needed to spend the UTXO used as the input for the assets (`8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:9` in my test) so that the associated single-use seal is closed to allow for the resulting state transition. However, this requirement is not mentioned in any of the examples that I have seen. If that is not true, then when and how exactly would that single-use seal be closed?

2. Even though I have specified a change tx output (`8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:0`), after the transfer was complete, that allocation was not showing in neither RGB nodes. That does not seem to be correct to me. Is it?

**Comment by dr-orlovsky** [comment]
Thank you very much for such detailed report and analysis! It's always **a big** pleasure to see somebody playing with the stuff you spent few years to create.

1. You are correct. However to understand how this work with the information you provide it's important to understand that RGB Node **is not a wallet** (even read-only) and can't report balances or anything like that. It's only role is to verify and hold (and never loose) client-validated data. So, it's storage and cache is append-only, never modify or delete. It does not know or care which outputs are really exists, owned by you etc; it verifies that txid exists and is spend only when does validation (using electrum backend) and never verifies whether some output unspent.

    So, yo get it working for real transfers you will need a wallet, which will (1) query RGB Node for everything it knows about outputs with assets; (2) filter only those outputs which are known to wallet as the own, (3) determine which of them are unspent and basing on this will calculate balance/present history/create new PSBTs for spending the assets.

    Right now we are doing first wallet for RGB; WIP is here: https://github.com/mycitadel/mycitadel-node. Next week it will be filled with the code for working with RGB Node and doing the tasks I described above. Also, with all frameworks we have developed (for [bitcoin wallets](https://github.com/LNP-BP/descriptor-wallet) and for rapid application development, such as [Internet2](https://github.com/internet2-org/rust-internet2)) it will take few more weeks to get it working from command-line (we also have mobile version, but it's not public; however expected to be released to closed beta within a month).

2. I think this may be a bug. I will do the investigation.

**Comment by dr-orlovsky** [comment]
2. You paid the change back to the address that is spent by PSBT, that's why it is lost. The cache data remembers the initial amount and does not update it since there is no point: it's important for wallet history to remember spent amounts; and this change allocation "never happened" from the point of view of the client-validated reality so ignored by the node.

**Comment by dr-orlovsky** [comment]
BTW, yesterday we released RGB Node v0.3 - you can try that version. The command-line API has not changed since v0.2

**Comment by claudiosdc** [comment]
> 1. You paid the change back to the tx output that contained initial issuance and was spent by PSBT, that's why it is lost. The cache remembers the initial amount issues to that TxO and does not update it since there is no point: it's important for wallet's history to remember and report previously held and spent amounts; and this change allocation "never happened" from the point of view of the client-validated reality so ignored by the node.

@dr-orlovsky, I'm afraid you are mistaken. Those two tx outputs are different, even though they share the same bitcoin transaction. The initial issuance of the asset was assigned to tx output `8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:9` (`vout: 9`), and that was the transaction output passed along with the `-i` option to the `fungible transfer` command. On the other hand, the transaction output chosen to receive the asset change back, and passed along with the `-a` option to the `fungible transfer` command was `8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:0` (`vout`: 0).

**Comment by claudiosdc** [comment]
Regarding the transfer change issue, by looking at the consignment data generated by the `fungible transfer` command (see below), one will notice that the transfer of the change amount to the designated tx output (`8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607:0`) is shown in the state transition (`owned_rights` field of `Transition` structure).

```shell
 Consignment { 
  version: 0, 
  genesis: Genesis { 
   schema_id: 86aa391b51d5eecb567a207a76eaae0e21307b22245466297a586045929e1b16, 
   chain: Testnet3, 
   metadata: Metadata(
    {
     0: {String("XYZB")}, 
     1: {String("XYZ Test 2")}, 
     2: {String("Test RGB asset #2")}, 
     3: {U8(2)}, 
     4: {I64(1611423235)}, 
     160: {U64(3000)}
    }
   ), 
   owned_rights: {
    161: DiscreteFiniteField(
     [
      Revealed { 
       seal_definition: TxOutpoint(
        OutpointReveal { 
         blinding: 4828134903816463740, 
         txid: 8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607, 
         vout: 9 
        }
       ), 
       assigned_state: Revealed { 
        value: 3000, 
        blinding: SecretKey(0000000000000000000000000000000000000000000000000000000000000001) 
       } 
      }
     ]
    )
   }, 
   public_rights: {}, 
   script: [] 
  }, 
  endpoints: [
   (
    c08ecd6ad4b7a7f98ef39c567a50f711c217b1aa1cd656cd561eb90c62d9a540, 
    aa4065bcbe3cd4afb1ea69dad5fa4f4e420676a5ac7a53f75d0a6c0e918ad515
   )
  ], 
  state_transitions: [
   (
    Anchor { 
     txid: abcd44508def1b27880adafd330c5c756ffb76b5f9c920b3b5f62d1a613d9823, 
     commitment: MultimsgCommitment { 
      commitments: [
       MultimsgCommitmentItem { 
        protocol: Some(1432bfb6f98206ac49dd3726c3b13faa23f9c5271058a50ffbf5308b7d85adc2), 
        commitment: 40a5d9620cb91e56cd56d61caab117c211f7507a569cf38ef9a7b7d46acd8ec0 
       }, 
       MultimsgCommitmentItem { 
        protocol: None, 
        commitment: 9aabd7e7a9aa70c53731c305d05ca427697b36b0a3e01246390481118460bdc2 
       }, 
       MultimsgCommitmentItem { 
        protocol: None, 
        commitment: ff57ed2ce97970cb1d4267e6016192c8bda8d1d683746e7edee58e974981d40d 
       }
      ], 
      entropy: Some(4589736911286851622) 
     }, 
     proof: Proof { 
      pubkey: PublicKey(
       2b8ba7da2e1b3ba211f2cb80dfabcde8c24c28206f1e62d3d1ff763845c94ca5e64648e3f07aba8c9054295c1285202a82f18e323714a1e6a01c378319da42d2
      ), 
      source: SinglePubkey 
     } 
    }, 
    Transition { 
     transition_type: 0, 
     metadata: Metadata({}), 
     parent_owned_rights: {
      c2ad857d8b30f5fb0fa5581027c5f923aa3fb1c32637dd49ac0682f9b6bf3214: {
       161: [0]
      }
     }, 
     owned_rights: {
      161: DiscreteFiniteField(
       [
        ConfidentialAmount { 
         seal_definition: TxOutpoint(
          OutpointReveal { 
           blinding: 9250875418385136544, 
           txid: 8d900009bdec06e1e4e46a91de70d6d94e5d081e017adafaf88e185bed09e607, 
           vout: 0 
          }
         ), 
         assigned_state: Confidential { 
          commitment: Commitment(
           09c623d6b45090a670c2b03ee54cfdbe9a93ff4dd6885203c67e9f357dac2d2678
          ), 
          bulletproof: RangeProof(
           72069ff636e10eb2837f7be7636110935891220e6d0cbccc02a6bb1d5d33602773215f11c0f0fc7f8cbd43bb7885be5ef3d1d9e78d1b7cd8f54458dbaafc28680179472f1cf7ab3fa9fc56c29d137195b7bb74b5879dadcc23e240e27ed1160f387dbd5ec5439318ed1320a996bf25ff18620553b562f8770ec82f0d641d78d3304a28cdbedecf72e925f70938740d9a3fa18387e329be71459c2fd0412e09c952ada7b14c0225ee57098f12b09f98088480883da81554c68a97bc915559a9ad10e183b02428629553e853058682d303d39ee42cba97c2073504dc170529b9e8d09c1d56ea85bcd6ca58a5e4d3644a5b9d2db950c85c3a9d13b918f454a09f7efb4fb11c945394fc96aa4604e9b55dc2d43de4f9929fc3150bb13f348a0ae121927faa8ba96236c7aa84b90bb09138c9b37635d003cf85644ff0c80eb6979cf0420c102732c66a12c34bbdfacd6d094e89aa4f450b4d7563416ee9ead01564284e6202ab732688c685e75f2fd52c2b62d940db938343e5f19b7e4b3a4daa2c3f41b5ca3fb46bf6a7445d7853a2679cc4e743c15254c2f55158bb0a616eb9625c39caec70dbc0dbe171a237e78467a2a8ee1f4ce0691f6a82d627a875d0a69fc6e3328c67b8a1fd3128111a02229d6cef3e6b051fa52edfa08d1d4b641c6d0b78daad4a4ee096267991d5e60fe117c2466e1c7189897608c49f02e3e9db12264b3ace078d08d7c65a9d6a01dd9fbda6cff9479535dcb9673a5d4a65819eecc26115f7d39426111414c79cbe4dbdd90a3d307427971d7fe10fa3c101f3b37ffadd0fd2da4ed1bd6e4a105a82ed7169f17bf04b3a54eaf44c8f5020f2e2fca4f33c57e9d807123ce381ffbd843817fffbb1617dd82241c4b322848a5e8b671225a040527bf50584c27fa4c5ec904f8f93baf3176b3787ac6ceb1e0830a2ab13eeb44066a7
          )[675] 
         } 
        }, 
        ConfidentialSeal { 
         seal_definition: aa4065bcbe3cd4afb1ea69dad5fa4f4e420676a5ac7a53f75d0a6c0e918ad515, 
         assigned_state: Revealed { 
          value: 30, 
          blinding: SecretKey(2d7b929dfa04dcbeb9db18273ce3a60633e26a5355b610c4fb5de0fa9a99a373) 
         } 
        }
       ]
      )
     }, 
     public_rights: {}, 
     script: [] 
    }
   )
  ], 
  state_extensions: [] 
 }
```

However, after reviewing the RGB node source code, I have noticed that the `known_allocations` field of the `Asset` structure maintained in the local cache is only updated (new entries added) during the processing of the `fungible accept` command. This explains why the change allocation is not added to the asset status at the sending RGB node. That does not seem to be right to me though. I would expect the local asset status to be updated during the processing of the `fungible transfer` asset.

**Comment by claudiosdc** [comment]
I have opened a new issue (#129) reporting the issue with the asset change.

**Comment by dr-orlovsky** [comment]
The sending party needs to do accept procedure separately of transfer operation, otherwise, if the corresponding witness transaction do not get mined the append-only stash will never see the asset as not being spent.

**Comment by dr-orlovsky** [comment]
Have we closed all questions unrelated to #129 here?

**Comment by claudiosdc** [comment]
> Have we closed all questions unrelated to #129 here?

Yes, we have. Thank you.
